Labs/11. Games/code-examples/canvas-game/src/Game.js

```diff
@@ -1,251 +1,371 @@
-            this.#ball.dx = -this.#ball.dx;
-        }
+  }
 
-        //Bouncing off the top
-        if (this.#ball.y + this.#ball.dy < this.ballRadius) {
-            this.#ball.dy = -this.#ball.dy;
-        }
+  drawScore() {
+    this.#context.font = "16px Arial";
+    this.#context.fillStyle = "#000000";
+    this.#context.fillText("Score: " + this.score, 8, 20);
+  }
 
-        //Bouncing off the bottom
-        else if (this.#ball.y + this.#ball.dy > this.#canvas.height - this.ballRadius) {
+  drawLives() {
+    this.#context.font = "16px Arial";
+    this.#context.fillStyle = "#000000";
+    this.#context.fillText("Lives: " + this.lives, this.#canvas.width - 65, 20);
+  }
 
-            if (this.#ball.x > this.#paddle.x && this.#ball.x < this.#paddle.x + this.paddleWidth) {
-                this.#ball.dy = -this.#ball.dy;
-            }
-            else {
-                this.lives--;
-                if (this.lives === 0) {
-                    alert('GAME OVER');
-                    document.location.reload();
-                }
-                else {
-                    this.#ball.x = this.#canvas.width / 2;
-                    this.#ball.y = this.#canvas.height - 30;
-                    this.#ball.dx = this.#ball.dxDefault;
-                    this.#ball.dy = this.#ball.dyDefault;
-                    //center the paddle
-                    this.#paddle.x = (this.#canvas.width - this.paddleWidth) / 2;
-                }
+  drawFPS() {
+    this.#context.font = "16px Arial";
+    this.#context.fillStyle = "#000000";
+    this.#context.fillText(
+      `FPS: ${Math.round(this.#currentFPS)} (${this.#simulatedRefreshRate}Hz)`,
+      8,
+      40
+    );
+  }
+
+  setSimulatedRefreshRate(hz) {
+    this.#simulatedRefreshRate = hz;
+    this.#frameDelay = 1000 / hz;
+  }
+
+  collisionDetection() {
+    for (let r = 0; r < this.brickRowCount; r++) {
+      for (let c = 0; c < this.brickColumnCount; c++) {
+        const brick = this.#bricks[r][c];
+        if (brick.status === 1) {
+          if (
+            this.#ball.x > brick.x &&
+            this.#ball.x < brick.x + this.brickWidth &&
+            this.#ball.y > brick.y &&
+            this.#ball.y < brick.y + this.brickHeight
+          ) {
+            this.#ball.dy = -this.#ball.dy;
+            brick.status = 0;
+            this.score++;
+
+            if (this.score === this.brickRowCount * this.brickColumnCount) {
+              alert("YOU WIN, CONGRATS!");
+              document.location.reload();
             }
+          }
         }
+      }
+    }
+  }
 
-        //move the paddle right if the right control button is pressed
-        if (this.rightKeyPressed && this.#paddle.x < this.#canvas.width - this.paddleWidth) {
-            this.#paddle.x += 7;
-        }
-        //move the paddle left if the left control button is pressed
-        else if (this.leftKeyPressed && this.paddl.X > 0) {
-            this.#paddle.x -= 7;
-        }
+  update(dt) {
```

Comment:
Cred că ar fi bine dacă metoda ar avea un nume mai sugestiv, prin adaugarea a 1-2 cuvinte suplimentare la nume in afara de update.